go/runtime/scheduling/simple/txpool/priorityqueue/priority_queue.go

```diff
@@ -0,0 +1,234 @@
+	"github.com/oasisprotocol/oasis-core/go/runtime/scheduling/simple/txpool/api"
+	"github.com/oasisprotocol/oasis-core/go/runtime/transaction"
+)
+
+// Name is the name of the tx pool implementation.
+const Name = "priority-queue"
+
+type item struct {
+	tx *transaction.CheckedTransaction
+}
+
+func (i item) Less(other btree.Item) bool {
+	i2 := other.(*item)
+	if p1, p2 := i.tx.Priority(), i2.tx.Priority(); p1 != p2 {
+		return p1 > p2
+	}
+	// If transactions have same priority, sort arbitrary.
+	h1 := i.tx.Hash()
+	h2 := i2.tx.Hash()
+	return bytes.Compare(h1[:], h2[:]) < 0
+}
+
+type priorityQueue struct {
+	sync.Mutex
+
+	priorityIndex *btree.BTree
+	transactions  map[hash.Hash]*item
+
+	maxTxPoolSize uint64
+
+	poolWeights  map[string]uint64
+	weightLimits map[string]uint64
+}
+
+// Implements api.TxPool.
+func (q *priorityQueue) Name() string {
+	return Name
+}
+
+// Implements api.TxPool.
+func (q *priorityQueue) Add(tx *transaction.CheckedTransaction) error {
+	q.Lock()
+	defer q.Unlock()
+
+	// Check if there is room in the queue.
+	if uint64(q.priorityIndex.Len()) >= q.maxTxPoolSize {
+		return api.ErrFull
+	}
+
+	if err := q.checkTxLocked(tx); err != nil {
+		return err
+	}
+
+	item := &item{tx: tx}
+	q.priorityIndex.ReplaceOrInsert(item)
+	q.transactions[tx.Hash()] = item
+	for k, v := range tx.Weights() {
+		q.poolWeights[k] += v
+	}
+
+	return nil
+}
+
+// Implements api.TxPool.
+func (q *priorityQueue) GetBatch(force bool) []*transaction.CheckedTransaction {
+	q.Lock()
+	defer q.Unlock()
+
+	// Check if a batch is ready.
+	var weightLimitReached bool
+	for k, v := range q.weightLimits {
+		if q.poolWeights[k] >= v {
+			weightLimitReached = true
+			break
+		}
+	}
+	if !weightLimitReached && !force {
+		return nil
+	}
+
+	var batch []*transaction.CheckedTransaction
+	batchWeights := make(map[string]uint64)
+	for w := range q.weightLimits {
+		batchWeights[w] = 0
+	}
+	toRemove := []*item{}
+	q.priorityIndex.Ascend(func(i btree.Item) bool {
+		item := i.(*item)
+
+		// Check if the call fits into the batch.
+		// XXX: potentially there could be smaller transactions that would
+		// fit, which this will miss. Could do some lookahead.
+		for w, limit := range q.weightLimits {
+			batchWeight := batchWeights[w]
+
+			txW := item.tx.Weight(w)
+			// Transaction weight greater than the limit. Drop the tx from the pool.
+			if txW > limit {
+				toRemove = append(toRemove, item)
+				return true
+			}
+
+			// Batch full, schedule the batch.
+			if batchWeight+txW > limit {
+				return false
+			}
+		}
+
+		// Add the tx to the batch.
+		batch = append(batch, item.tx)
+		for w, val := range item.tx.Weights() {
+			if _, ok := batchWeights[w]; ok {
+				batchWeights[w] += val
+			}
+		}
+
+		return true
+	})
+
+	// Remove transactions discovered to be too big to even fit the batch.
+	// This can happen if weight limits changed after the transaction was
+	// already set to be scheduled.
+	for _, tw := range toRemove {
```

Comment:
This now needs to also update the pool weights.

Reply:
that's right, i should also add a test case for this

Reply:
fixed, thanks!